src/freertos_drivers/common/PCA9685PWM.hxx

```diff
@@ -0,0 +1,414 @@
+
+        register_write(address, rd_data);
+    }
+
+    /// Set the pwm duty cycle
+    /// @param channel channel index (0 through 15)
+    /// @param counts counts for PWM duty cycle
+    void set_pwm_duty(unsigned channel, uint16_t counts)
+    {
+        HASSERT(channel < NUM_CHANNELS);
+        duty_[channel] = counts;
+
+        portENTER_CRITICAL();
+        dirty_ |= 0x1 << channel;
+        portEXIT_CRITICAL();
+
+        sem_.post();
+    }
+
+    /// Get the pwm duty cycle
+    /// @param channel channel index (0 through 15)
+    /// @return counts for PWM duty cycle
+    uint16_t get_pwm_duty(unsigned channel)
+    {
+        HASSERT(channel < NUM_CHANNELS);
+        return duty_[channel];
+    }
+
+    /// Set the pwm duty cycle
+    /// @param channel channel index (0 through 15)
+    /// @param counts counts for PWM duty cycle
+    void write_pwm_duty(unsigned channel, uint16_t counts)
+    {
+        LedCtl ctl;
+        if (counts >= MAX_PWM_COUNTS)
+        {
+            ctl.on.fullOn = 1;
+            ctl.off.fullOff = 0;
+        }
+        else if (counts == 0)
+        {
+            ctl.on.fullOn = 0;
+            ctl.off.fullOff = 1;
+        }
+        else
+        {
+            // the "256" count offset is to help average the current accross
+            // all 16 channels when the duty cycle is low
+            ctl.on.counts = (channel * 256);
+            ctl.off.counts = (counts + (channel * 256)) % 0x1000;
+        }
+
+        htole16(ctl.on.word);
+        htole16(ctl.off.word);
+
+        Registers offset = (Registers)(LED0_ON_L + (channel * 4));
+        register_write_multiple(offset, &ctl, sizeof(ctl));
+    }
+
+    /// Write to an I2C register.
+    /// @param address address to write to
+    /// @param data data to write
+    void register_write(Registers address, uint8_t data)
+    {
+        uint8_t payload[] = {address, data};
+        ::write(i2c_, payload, sizeof(payload));
+    }
+
+    /// Write to multiple sequential I2C registers.
+    /// @param address address to start write at
+    /// @param data array of data to write
+    /// @param count number of data registers to write in sequence
+    void register_write_multiple(Registers address, void *data, size_t count)
+    {
+        uint8_t payload[count + 1];
+        payload[0] = address;
+        memcpy(payload + 1, data, count);
+        ::write(i2c_, payload, sizeof(payload));
+    }
+
+    /// Allow access to private members
+    friend PCA9685PWMBit;
+
+    /// Wakeup for the thread processing
+    OSSem sem_;
+
+    /// I2C file descriptor
+    int i2c_;
+
+    /// local cache of the duty cycles
+    std::array<uint16_t, NUM_CHANNELS> duty_;
+
+    /// set if the duty_ value is updated (bit mask)
+    uint16_t dirty_;
+
+    /// I2C address of the device
+    uint8_t i2cAddress_;
+
+    DISALLOW_COPY_AND_ASSIGN(PCA9685PWM);
+};
+
+/// Specialization of the PWM abstract interface for the PCA9685
+class PCA9685PWMBit : public PWM
```

Comment:
I think it would be worth to check whether these instances can be created in .rodata memory. I don't quite remember now what was needed, but for the SerialGPIO we had to use this trick in Portand to fit the project into the 32k ram of the tiva.

Reply:
I just reviewed the tcs_io code, and I don't see anything different.  The instances do not appear to be in *.rodata.  We did make them a global array.